interfaces/devdoc/threadpool_requirements.md

```diff
@@ -31,6 +31,12 @@ MOCKABLE_FUNCTION(, THANDLE(THREADPOOL), threadpool_create, EXECUTION_ENGINE_HAN
 MOCKABLE_FUNCTION(, int, threadpool_open, THANDLE(THREADPOOL), threadpool);
 MOCKABLE_FUNCTION(, void, threadpool_close, THANDLE(THREADPOOL), threadpool);
 
+MOCKABLE_FUNCTION(, PVOID, threadpool_create_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_FUNCTION, work_function, PVOID, work_function_context);
```

Comment:
>PVOID [](http://example.com/codeflow?start=20&length=5)

And this should return the `THREADPOOL_WORK_ITEM_HANDLE`, which should be the argument type for the other two functions (instead of `PVOID`)

Reply:
We are returning the pointer to `WORK_ITEM_CONTEXT_V2` which contains `work_function`, `work_function_context`. and `ptp_work` and saving it in the Queue Processor or any caller for that matter. 
In either case WORK ITEM or WORK ITEM CONTEXT are not available in Linux, hence using PVOID for return and argument type.
`THREADPOOL_WORK_FUNCTION` is available in both WIN32 and LINUX

Reply:
So generally, our conventions are to provide a header with an opaque type and then functions to operate on it.

As in we have `typedef struct SOMETHING* SOMETHING_HANDLE`.

"users" of the module cannot create a "struct SOMETHING", they don't know anything about the data it stores. They just get a pointer (SOMETHING_HANDLE) and can pass that into the functions. This gives you the minimal type checking that C has (not the same level as C++ but better than nothing) and it encapsulates the implementation.

We should do the same thing for the work item. The interface should have some opaque type which is used for both windows and linux. The implementation can be different, just like the `struct THREADPOOL_TAG` we have now.